go-controller/pkg/ovn/master.go

```diff
@@ -382,19 +381,53 @@ func (oc *DefaultNetworkController) addNode(node *kapi.Node) ([]*net.IPNet, erro
 			v6Addr = ip
 		}
 	}
-	updatedNodeAnnotation, err := util.CreateNodeGatewayRouterLRPAddrAnnotation(nil, v4Addr, v6Addr)
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal node %q annotation for Gateway LRP IP %v",
-			node.Name, gwLRPIPs)
+
+	// Retry if it fails because of potential conflict which is transient. Return error in the
+	// case of other errors (say temporary API server down), and it will be taken care of by the
+	// retry mechanism.
+	resultErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// Informer cache should not be mutated, so get a copy of the object
+		node, err := oc.watchFactory.GetNode(nodeName)
+		if err != nil {
+			return err
+		}
+
+		cnode := node.DeepCopy()
+		cnode.Annotations, err = util.CreateNodeGatewayRouterLRPAddrAnnotation(cnode.Annotations, v4Addr, v6Addr)
+		if err != nil {
+			return fmt.Errorf("failed to marshal node %q annotation for Gateway LRP IP %v",
+				node.Name, gwLRPIPs)
+		}
+		return oc.kube.PatchNode(node, cnode)
+	})
+	if resultErr != nil {
+		return fmt.Errorf("failed to update node %s annotation", nodeName)
 	}
+	return nil
+}
 
-	hostSubnets, err := oc.allocateNodeSubnets(node, oc.masterSubnetAllocator)
+func (oc *DefaultNetworkController) addNode(node *kapi.Node) ([]*net.IPNet, error) {
+	hostSubnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
 	if err != nil {
+		// Log the error and try to allocate new subnets
+		klog.Infof("Failed to get node %s host subnets annotations: %v", node.Name, err)
 		return nil, err
 	}
 
-	hostSubnetsMap := map[string][]*net.IPNet{oc.GetNetworkName(): hostSubnets}
-	err = oc.UpdateNodeAnnotationWithRetry(node.Name, hostSubnetsMap, updatedNodeAnnotation)
+	// OVN can work in single-stack or dual-stack only.
+	currentHostSubnets := len(hostSubnets)
+	expectedHostSubnets := 1
+	// if dual-stack mode we expect one subnet per each IP family
+	if config.IPv4Mode && config.IPv6Mode {
+		expectedHostSubnets = 2
+	}
+
+	if expectedHostSubnets != currentHostSubnets {
+		klog.Errorf("Allocated Subnets %v on Node %s", hostSubnets, node.Name)
+		return nil, fmt.Errorf("failed to get expected host subnets for node : %s, expected subnets : %d current subnets: %d", node.Name, expectedHostSubnets, currentHostSubnets)
+	}
+
+	err = oc.updateNodeAnnotationWithRetry(node.Name)
```

Comment:
hmm I am a bit confused with the change of code here between adding a new `updateNodeAnnotationWithRetry` and now we will try to update this annotation from both cluster manager and ovnkube-master? .. maybe I need to come back to this in Round2 of reviews.

Reply:
Yes.  We would.  cluster-manager will update node subnets annotations and a few others (in the next PR) where ovnkube-master(s) would update other annotations.

Reply:
> Another potential problem here is applying Patch for annotations. AFAIK it is only safe for concurrent use for different annotation keys, so we need to make sure the requests for the same annotation key are synchronized, or we need to use update instead of patch.
> 
> It looks like cluster managers for default and secondary networks may update `k8s.ovn.org/node-subnets` annotations at the same time?

This could also happen with the present ovnkube-master too right ?  i.e DefaultNetworkController nand SecondaryLayer3Controller updating the node-subnets annotations at the same time ?

Please correct me if I'm wrong.

Reply:
In the next revision,  I'll use kube.UpdateNode() instead of PatchNode().

In my local testing I found once that a node was missing its subnet annotation for the secondary network even though the relevant code was executed to UpdateNodeAnnotationWithRetry() was called.

Reply:
Yes, looks like this Patch conflict should be also fixed for secondary controllers, let's see if I can get this https://github.com/ovn-org/ovn-kubernetes/pull/3405 merged soon, and then we can generalize update method for gateway router annotation